### ISSUE TITLE
add support for adaptive softmax to reduce memory

### ISSUE THREAD

**Issue by ibeltagy** [bigscience-workshop/Megatron-DeepSpeed#166]
This:  https://pytorch.org/docs/stable/generated/torch.nn.AdaptiveLogSoftmaxWithLoss.html
Why: reduce memory requirements with a large vocabulary. Relevant for the multilingual experiments.

**Comment by stas00** [comment]
Gentleman, I have a silly question - I am missing the context - and it appears that you propose to switch to  `AdaptiveLogSoftmaxWithLoss` to use less memory? Where are you dealing with OOM in the first place?

I ask since I don't see anywhere in the paper or the pytorch docs that it says that it reduces memory usage. It discusses speed but not memory. I suppose by the nature of the approach it probably should be more efficient in memory usage as well as since it runs faster it must be using less memory?

Thank you!

**Comment by thomasw21** [comment]
@jaketae so self.specs is how we build a PP version of GPT2, ie deepspeed supports lists of blocks. Concerning the upcasting, I'd say that it is necessary for softmax computation over the vocabulary (which is 50k big in english and 250k big in multi lingual). So I don't think removing the upcasting would be possible. TBD.

**Comment by stas00** [comment]
We haven't added it yet, as I'm trying a different solution first as explained in my previous comment.

But you can also try other MBS between 8 and 1 :)

**Comment by stas00** [comment]
Update:

So I'm still waiting to hear from Shaden about the ZeRO-1 issue, but you can already use the following solution, except you have to turn ZeRO-1 off in the config file (just set it to 0):

```
diff --git a/megatron/model/gpt_model.py b/megatron/model/gpt_model.py
index f0bb858..8af832e 100644
--- a/megatron/model/gpt_model.py
+++ b/megatron/model/gpt_model.py
@@ -280,4 +280,4 @@ class GPTModelPipe(PipelineModule,MegatronModule):
                          loss_fn=CrossEntropy,
                          topology=topo,
                          activation_checkpoint_interval=interval,
-                         partition_method='type:transformer')
\ No newline at end of file
+                         partition_method='type:embed|transformer')
```

This makes the embed layer on par with transformer layer when it performs the partitioning, so now it can have a whole stage for itself.

Please let me know if you still would like me to look into the adaptive softmax?

**Comment by stas00** [comment]
With TP=4 PP=32 

So with the above change and lowering n_layers to -2, I get this new map:

```
     0: _to_float16
     1: EmbeddingPipe
     2: <lambda>
     3: ParallelTransformerLayerPipe
stage=1 layers=2
     4: ParallelTransformerLayerPipe
     5: ParallelTransformerLayerPipe
[...]
stage=30 layers=2
    62: ParallelTransformerLayerPipe
    63: ParallelTransformerLayerPipe
stage=31 layers=5
    64: ParallelTransformerLayerPipe
    65: <lambda>
    66: MixedFusedLayerNorm
    67: EmbeddingPipe
    68: float16_to_fp32
  loss: CrossEntropy
```

Note, we had to chop off 2 layers!

So now `EmbeddingPipe` and `ParallelTransformerLayerPipe` are on equal grounds. The former is about 2x larger than the latter, so the first and last stages where they share it. are 1.5x larger than stages with 2 `ParallelTransformerLayerPipe` 
 
If need be we could further re-tune the partitioning to give `EmbeddingPipe` more space - actually we will most likely want to do that to equalize the weights and utilize the gpus better. So a whole stage to  `EmbeddingPipe` (2 times) and the rest is pairs of `ParallelTransformerLayerPipe`.

Memory usage of the first gpu is only 24GB now, the middle ones is only 18GB, so it's not very balanced, but as you can see there is plenty of free gpu memory to go around.

we will need to hack https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/pipe/module.py#L378-L384 to support `partition_method` `type:embed:2|transformer:1` - or something like that - now the embed weights will get 2x partitioning weights and will get its own stage and all stages will be more balanced.

And we will need to chop of 4 layers in the config then. So instead of 64 - it'll be 60. 

----

Now, because of the bug in Z1 https://github.com/microsoft/DeepSpeed/issues/1522 we lose the nice sharding of optimizer states over 4 gpus (TP=4), which is a huge loss. But by adding `--use-bnb-optimizer` we only use 1/4 of memory consumed by optimizer states, so we get exactly the same memory usage. If ZeRO1 gets fixed, there will be further saving.
 
We are in the process of experimenting with BNB to see that it produces similar results to normal Adam and thus can use it for our training.

I will update this shortly with more details and the script.

Here is the full slurm script:

[tr8-104B-mt5-32-30m.txt](https://github.com/bigscience-workshop/Megatron-DeepSpeed/files/7502004/tr8-104B-mt5-32-30m.txt)

Note that it requires the application of:
1. the diff in https://github.com/bigscience-workshop/Megatron-DeepSpeed/issues/166#issuecomment-961594612
2. cherry-picking https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/182
```
git cherry-pick 2f03484
git cherry-pick dab2064
```

**Comment by stas00** [comment]
Actually, partitioning by `parameters` instead of `partition_method='type:embed|transformer'` gives us the same result:

```
diff --git a/megatron/model/gpt_model.py b/megatron/model/gpt_model.py
index f0bb858..3a3a9db 100644
--- a/megatron/model/gpt_model.py
+++ b/megatron/model/gpt_model.py
@@ -280,4 +280,4 @@ class GPTModelPipe(PipelineModule,MegatronModule):
                          loss_fn=CrossEntropy,
                          topology=topo,
                          activation_checkpoint_interval=interval,
-                         partition_method='type:transformer')
+                         partition_method='parameters')
```
